tcheck_test.go

```diff
@@ -42,64 +41,128 @@ func healthOk(ctx thrift.Context) (ok bool, message string) {
+	noHandler := setupServer(t, nil)
+	defer noHandler.Close()
+
+	unhealthyHandler := setupServer(t, func(_ thrift.Context) (ok bool, msg string) {
+		return false, "test-error"
+	})
+	defer unhealthyHandler.Close()
+
+	healthyHandler := setupServer(t, func(_ thrift.Context) (ok bool, msg string) {
+		return true, ""
+	})
+
+	tests := []struct {
+		msg      string
+		peer     string
+		svc      string
+		fn       thrift.HealthFunc
+		wantExit int
+		wantErr  string
+	}{
+		{
+			msg:      "missing service",
+			peer:     "127.0.0.1",
+			svc:      "",
+			wantExit: _exitUsage,
+		},
+		{
+			msg:      "missing peer",
+			peer:     "",
+			svc:      "svc",
+			wantExit: _exitUsage,
+		},
+		{
+
+			msg:      "healthy server",
+			peer:     healthyHandler.PeerInfo().HostPort,
+			svc:      "svc",
+			wantExit: 0,
+		},
+		{
+			msg:      "no health handler",
+			peer:     noHandler.PeerInfo().HostPort,
+			svc:      "svc",
+			wantExit: _exitUnknownUnhealthy,
+			wantErr:  "ErrCodeBadRequest",
+		},
+		{
+			msg:      "unhealthy health handler",
+			peer:     unhealthyHandler.PeerInfo().HostPort,
+			svc:      "svc",
+			wantExit: _exitExplitiUnhealthy,
+			wantErr:  "test-error",
+		},
+	}
 
-	channel.Close()
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			err := healthCheck(tt.peer, tt.svc)
+			if tt.wantExit > 0 {
+				require.Error(t, err)
+				assert.Equal(t, tt.wantExit, getExitCode(err), "Unexpected error code")
+				assert.Contains(t, err.Error(), tt.wantErr, "Missing expected error")
+				return
+			}
+
+			require.NoError(t, err)
+		})
+	}
 }
 
-func TestNotOk(t *testing.T) {
-	channel, hostPort := SetupServer(t, healthNotOk)
-	strOut, err := Run([]string{fmt.Sprintf("--peer=%s", hostPort), "--serviceName=testing"})
-
-	strErr := fmt.Sprintf("%v", err)
-	assert.Equal(t, "exit status 3", strErr, "correct return code")
+func TestIntegrationSuccess(t *testing.T) {
+	server := setupServer(t, nil)
+	defer server.Close()
 
-	assert.Equal(t, "NOT OK hello world\n", strOut, "tcheck correct output")
+	// Set up a default health handler.
+	thrift.NewServer(server)
 
-	channel.Close()
+	os.Args = []string{"tcheck", "--peer", server.PeerInfo().HostPort, "--serviceName", server.ServiceName()}
+	main()
 }
 
-func TestNoHandler(t *testing.T) {
-	channel, hostPort := SetupServer(t, nil)
-	strOut, err := Run([]string{fmt.Sprintf("--peer=%s", hostPort), "--serviceName=testing"})
+func TestIntegrationError(t *testing.T) {
+	defer func() { _osExit = os.Exit }()
 
-	strErr := fmt.Sprintf("%v", err)
-	assert.Equal(t, "exit status 2", strErr, "correct return code")
-
-	errMsg := tchannel.NewSystemError(tchannel.ErrCodeBadRequest, "no handler for service")
-	expectedPrefix := "NOT OK testing\nError: " + errMsg.Error()
-	assert.True(t, strings.HasPrefix(strOut, expectedPrefix),
-		"Expected STDOUT to have prefix:\n%s\nbut got:\n%s", expectedPrefix, strOut)
+	var exitCode int
+	_osExit = func(code int) {
+		exitCode = code
```

Comment:
nit: `main()` expects `_osExit` to stop executing the function, otherwise it
will also print `OK` on failure. Since you're already calling `main` in a
goroutine, you can [`runtime.Goexit`] to kill the goroutine when `_osExit` is
called.

  [`runtime.Goexit`]: https://golang.org/pkg/runtime/#Goexit